pkg/assetattributor/asset_inventory_api.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	if asset.IP == "" && asset.Hostname == "" {
+		return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetNotFoundError{
+			Inner:          fmt.Errorf("asset has no IP or hostname"),
+			AssetID:        fmt.Sprintf("%d", asset.ID),
+			ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+			AssetInventory: assetInventoryAPIIdentifier,
+		}
+	}
+
+	// Since asset-inventory-api has APIs for asset lookup by both IP and hostname, make concurrent
+	// calls to asset-inventory-api's APIs for asset lookup by IP and hostname, wait for the calls
+	// to complete, and aggregrate the responses and errors onto buffered channels.
+	assetDetailsChan := make(chan domain.CloudAssetDetails, 2)
+	errChan := make(chan error, 2)
+	wg := &sync.WaitGroup{}
+
+	if asset.IP != "" {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			assetDetails, err := n.fetchAsset(ctx, resourcePathTypeIP, asset.IP, asset.LastScanned)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			assetDetailsChan <- assetDetails[0]
+		}()
+	}
+
+	if asset.Hostname != "" {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			assetDetails, err := n.fetchAsset(ctx, resourcePathTypeHostname, asset.Hostname, asset.LastScanned)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			assetDetailsChan <- assetDetails[0]
+		}()
+	}
+	wg.Wait()
+
+	// Once all goroutines have completed and all errors received, close the error channel and
+	// range over the buffered output, exiting on any fatal errors.
+	close(errChan)
+	outerErrs := []error{}
+	for e := range errChan {
+		switch e.(type) {
+		case httpMultipleAssetsFoundError:
+			return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetInventoryMultipleAssetsFoundError{
+				Inner:          e,
+				AssetID:        fmt.Sprintf("%d", asset.ID),
+				ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+				AssetInventory: assetInventoryAPIIdentifier,
+			}
+		case httpBadRequest:
+			return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetInventoryRequestError{
+				Inner:          e,
+				AssetID:        fmt.Sprintf("%d", asset.ID),
+				ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+				AssetInventory: assetInventoryAPIIdentifier,
+				Code:           http.StatusBadRequest,
+			}
+		}
+		outerErrs = append(outerErrs, e)
+	}
+
+	// Exit with an AssetNotFoundError error if both API calls returned non-fatal errors.
+	if len(outerErrs) == 2 {
+		return domain.NexposeAttributedAssetVulnerabilities{}, domain.AssetNotFoundError{
+			Inner:          combinedError{Errors: outerErrs},
+			AssetID:        fmt.Sprintf("%d", asset.ID),
+			ScanTimestamp:  asset.LastScanned.Format(time.RFC3339Nano),
+			AssetInventory: assetInventoryAPIIdentifier,
+		}
+	}
+
+	// Close the assetDetailsChan channel and append the buffered output to a slice.
+	close(assetDetailsChan)
+	assetDetails := []domain.CloudAssetDetails{}
+	for assetDetail := range assetDetailsChan {
```

Comment:
questions: since we only care about the first one, why add both to the slice?

Reply:
No real reason other than I thought this looked clean. I'm open to other implementations.